File: web/src/core/prayerType.ts
export enum PrayerType {
  Subuh = "subuh",
  Dzuhur = "dzuhur",
  Ashar = "ashar",
  Maghrib = "maghrib",
  Isya = "isya",
  Unknown = "unknown",
}

export function prayerDisplayName(p: PrayerType): string {
  switch (p) {
    case PrayerType.Subuh:
      return "Subuh";
    case PrayerType.Dzuhur:
      return "Dzuhur";
    case PrayerType.Ashar:
      return "Ashar";
    case PrayerType.Maghrib:
      return "Maghrib";
    case PrayerType.Isya:
      return "Isya";
    default:
      return "Tidak diketahui";
  }
}

export interface PrayerWindow {
  prayer: PrayerType;
  startMinute: number;
  endMinute: number;
}

/** Half-open windows [start, end) in minutes-from-midnight mapping wall-clock
 *  time to the active prayer. The app can replace this with real times from a
 *  prayer-time library; this fixed set is the offline fallback. */
export class PrayerWindows {
  constructor(public windows: PrayerWindow[]) {}

  static fallback(): PrayerWindows {
    return new PrayerWindows([
      { prayer: PrayerType.Subuh, startMinute: 4 * 60, endMinute: 6 * 60 },
      { prayer: PrayerType.Dzuhur, startMinute: 11 * 60 + 30, endMinute: 15 * 60 },
      { prayer: PrayerType.Ashar, startMinute: 15 * 60, endMinute: 18 * 60 },
      { prayer: PrayerType.Maghrib, startMinute: 18 * 60, endMinute: 19 * 60 },
      { prayer: PrayerType.Isya, startMinute: 19 * 60, endMinute: 24 * 60 },
    ]);
  }

  prayerAtMinutes(m: number): PrayerType {
    for (const w of this.windows) {
      if (m >= w.startMinute && m < w.endMinute) return w.prayer;
    }
    return PrayerType.Unknown;
  }
}

File: Sources/RakaatCore/PrayerType.swift
import Foundation

/// The five daily fardh prayers (plus unknown for ambiguous cases).
public enum PrayerType: String, Equatable, Sendable, CaseIterable {
    case subuh
    case dzuhur
    case ashar
    case maghrib
    case isya
    case unknown

    public var displayName: String {
        switch self {
        case .subuh: return "Subuh"
        case .dzuhur: return "Dzuhur"
        case .ashar: return "Ashar"
        case .maghrib: return "Maghrib"
        case .isya: return "Isya"
        case .unknown: return "Tidak diketahui"
        }
    }

    /// Canonical fardh rakaat count.
    public var rakaat: Int {
        switch self {
        case .subuh: return 2
        case .maghrib: return 3
        case .dzuhur, .ashar, .isya: return 4
        case .unknown: return 0
        }
    }
}

/// Half-open time windows [start, end) in minutes-from-midnight that map a
/// wall-clock time to the currently-active prayer. The iOS app fills these from
/// the Adhan library (location-based); this struct keeps inference testable and
/// provides a rough fallback.
public struct PrayerWindows: Sendable {
    /// (prayer, startMinute, endMinute). Order does not matter.
    public var windows: [(PrayerType, Int, Int)]

    public init(windows: [(PrayerType, Int, Int)]) {
        self.windows = windows
    }

    /// Rough fixed fallback when no location is available.
    public static let fallback = PrayerWindows(windows: [
        (.subuh,   4 * 60,  6 * 60),       // 04:00–06:00
        (.dzuhur, 11 * 60 + 30, 15 * 60),  // 11:30–15:00
        (.ashar,  15 * 60, 18 * 60),       // 15:00–18:00
        (.maghrib,18 * 60, 19 * 60),       // 18:00–19:00
        (.isya,   19 * 60, 24 * 60),       // 19:00–24:00
    ])

    public func prayer(atMinutes m: Int) -> PrayerType {
        for (p, s, e) in windows where m >= s && m < e { return p }
        return .unknown
    }
}
